src/NuGet.Core/NuGet.Protocol/Exceptions/QuotaExceededException.cs

```
@@ -0,0 +1,57 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.Runtime.Serialization;
+
+namespace NuGet.Protocol.Core.Types
+{
+    /// <summary>
+    /// Exception thrown when a quota limit is exceeded (HTTP 403 Forbidden with Retry-After header).
+    /// </summary>
+    [Serializable]
+    public class QuotaExceededException : FatalProtocolException
```

I see this exception is only used for push, not other commands like search or restore. But the HTTP error is 403 Forbidden. That's more likely to be access denied than some kind of quota error. If we create a new exception, I  think it should have a name that's less risk of being misleading.

---

Do we even need a separate exception?

---

The customer is asking for the http response's Retry-After header data to be available. But [the base Exception class has a generic `Data` dictionary](https://learn.microsoft.com/en-us/dotnet/api/system.exception.data?view=net-9.0), so we could throw any existing exception type and copy the values in there instead.

I think an ugly workaround without needing any NuGet changes is for the customer to override SourceRepository's default HttpClientHandler, and inject their own that has middleware to capture the Retry-After header. But that's quite a lot more lines of code than just catching an exception.

---

Yeah, I like reusing the Data dict. 

Not sure we'd need to define exceptions for every scenario.

---

What do you think about this overall approach:

* In the code that this PR currently creates instances of this QuoteExceededExeption, instead add the RetryAfter property to the existing HttpResponseException, then rethrow the exception so that we're not changing the callstack or nesting more exceptions with inner exceptions
* In FatalProtocolException's constructor, when an inner exception is provided, check if it has a RetryAfter property and if so, copy it into the current instance's Data dictionary.

---

Sure, I don't mind that.